### PR TITLE
cffi_message: Added scripts for compile message c functions

### PR DESCRIPTION
**Pull request by SudhanAruna**
Signed-off-by: Arunagirinadan Sudharshan <sudhan.aruna.97@gmail.com>

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -77,26 +77,30 @@ jobs:
...
       uses: actions/upload-artifact@v2
       with:
         name: python-pkgs
         path: ${{env.BuildDir}}/python/pkgs
+
     - id: ctest
       name: ctest
       working-directory: ${{env.BuildDir}}
-      run: ctest -C ${BuildType} -V -T Test --no-compress-output -R python3-test ${{matrix.ctest_extra}}
+      run: ctest -C ${BuildType} -V -T Test --no-compress-output -R proton_tests ${{matrix.ctest_extra}}
```
Comment (jiridanek) [inline .github/workflows/build.yml]: Github actions:

```
Done constructing a list of tests
No tests were found!!!
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -77,26 +77,30 @@ jobs:
...
       uses: actions/upload-artifact@v2
       with:
         name: Test_Results_${{matrix.os}}_${{matrix.buildType}}
         path: ${{env.BuildDir}}/Testing/**/*.xml
+
     - name: Upload Python & C build directories on failure
       uses: actions/upload-artifact@v2
       if: failure()
-      with:
+      with:
```
Comment (jiridanek) [inline .github/workflows/build.yml]: ;)

File: python/CMakeLists.txt [hunk 3]
```diff
@@ -17,34 +17,30 @@
...
+if (WIN32)
+  set(python_suffix .pyd)
+else ()
+  set(python_suffix .so)
+endif()
 
-swig_add_library(cproton LANGUAGE python SOURCES cproton.i)
-swig_link_libraries(cproton ${BINDING_DEPS} ${Python_LIBRARIES} -lm)
-set_target_properties(${SWIG_MODULE_cproton_REAL_NAME}
+add_library(_proton_core SHARED ${CMAKE_CURRENT_BINARY_DIR}/_proton_core.c)
```
Comment (jiridanek) [inline python/CMakeLists.txt:33]: Appveyor:

```
CMake Error at C:/Tools/vcpkg/scripts/buildsystems/vcpkg.cmake:702 (_add_library):
  _add_library cannot create target "_proton_core" because another target
  with the same name already exists.  The existing target is a shared library
  created in source directory "C:/projects/qpid-proton/python3".  See
  documentation for policy CMP0002 for more details.
Call Stack (most recent call first):
  python/CMakeLists.txt:33 (add_library)
CMake Error at python/CMakeLists.txt:35 (target_link_libraries):
  Attempt to add link library "qpid-proton-core" to target "_proton_core"
  which is not built in this directory.
  This is allowed only when policy CMP0079 is set to NEW.
```

File: python/proton/_common.py [hunk 4]
```diff
@@ -80,12 +82,16 @@ def unicode2utf8(string):
...
     elif isinstance(string, unicode):
         # py2 unicode, py3 str (via hack definition)
         return string
     elif isinstance(string, bytes):
         # py2 str (via hack definition), py3 bytes
         return string.decode('utf8')
-    raise TypeError("Unrecognized string type")
+    elif isinstance(string, ffi.CData):
+        # ffi.string return bytes for cdata type <check this comment pleaseeeeeeeeeeeeeeeeeeee>
+        return ffi.string(string).decode()
```
Comment (jiridanek) [inline python/proton/_common.py:95]: lgtm, that's what the docs says that happens, on Python3

File: python/proton_cffi_builder.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+
+    const char *pn_error_text(pn_error_t *error);
+
+    int pn_error_copy(pn_error_t *error, pn_error_t *src);
+
+
+"""
+
+
+ffi_builder.cdef(  type_h + object_h + error_h + codec_t +  message_h)
```
Comment (jiridanek) [inline python/proton_cffi_builder.py]: you've mentioned you have a script to generate all the cdef strings. please add it to the PR. It does not matter if it's not yet completely finished, or completely polished.

File: python/proton/_message.py [hunk 6]
```diff
@@ -446,7 +462,7 @@ def _get_reply_to_group_id(self):
         return utf82unicode(pn_message_get_reply_to_group_id(self._msg))
 
     def _set_reply_to_group_id(self, value):
-        self._check(pn_message_set_reply_to_group_id(self._msg, unicode2utf8(value)))
+        self._check(pn_message_set_reply_to_group_id(self._msg, unicode2utf8(value).encode()))
```
Comment (jiridanek) [inline python/proton/_message.py:465]: I want you to fix the CI issues first. But then, I have also this:

Doing `unicode2utf8(value).encode()` looks cumbersome. First, you convert `value` to `str`, then you convert it to `bytes`. It would make sense to me to replace the `unicode2utf8` usage here with a new function, which would directly return the value converted to `bytes` , and maybe it will also pass through a cffi `char *` values.

**Review by jiridanek — CHANGES_REQUESTED** [review]
Please fix the CI. I want to see a successful compilation in all CIs and I want to see the message tests running in CI.

File: python/proton/_common.py [hunk 7]
```diff
@@ -80,12 +82,16 @@ def unicode2utf8(string):
 
 def utf82unicode(string):
     """Convert C strings returned from proton-c into python unicode"""
-    if string is None:
+    if string is None or string == ffi.NULL:
         return None
     elif isinstance(string, unicode):
         # py2 unicode, py3 str (via hack definition)
         return string
```
Comment (jiridanek) [inline python/proton/_common.py:89]: something you'd be able to revisit in another PR down the road; the minimal Python version you need to support is Python 3.6, so any hacks for Python 2 should be deleted from the code

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -67,7 +67,7 @@ jobs:
         brew install libuv swig pkgconfig jsoncpp
     - name: cmake configure
       working-directory: ${{env.BuildDir}}
-      run: cmake "${{github.workspace}}" "-DCMAKE_BUILD_TYPE=${BuildType}" "-DCMAKE_INSTALL_PREFIX=${InstallPrefix}" -DRUNTIME_CHECK=${{matrix.runtimeCheck}} -DBUILD_BINDINGS=python3 ${{matrix.cmake_extra}}
+      run: cmake "${{github.workspace}}" "-DCMAKE_BUILD_TYPE=${BuildType}" "-DCMAKE_INSTALL_PREFIX=${InstallPrefix}" -DRUNTIME_CHECK=${{matrix.runtimeCheck}} -DBUILD_BINDINGS=python,python3 ${{matrix.cmake_extra}}
```
Comment (jiridanek) [inline .github/workflows/build.yml]:
```suggestion
      run: cmake "${{github.workspace}}" "-DCMAKE_BUILD_TYPE=${BuildType}" "-DCMAKE_INSTALL_PREFIX=${InstallPrefix}" -DRUNTIME_CHECK=${{matrix.runtimeCheck}} -DBUILD_BINDINGS="python;python3" ${{matrix.cmake_extra}}
```

and it has to be a semicolon, which is a bit tricky, because semicolon is a command separator in linux shell, therefore I also added quotes, to be sure I don't end up executing two shell commands separated by the semicolon...

this is documented in a comment in CMakeLists.txt, but nowhere else I could find; so I guess that's something to mention in the jira too... feature needs 1) clear error when you specify binding that does not exist, 2) documentation

Comment (jiridanek) [reply .github/workflows/build.yml]: @SudhanAruna this is the reasons no tests are running

Comment (SudhanAruna) [reply .github/workflows/build.yml]: > @SudhanAruna You still did not fix this. And please create a jira for the problem, as I wrote before.

Sorry I missed this notification regarding this comment.  By "create a jira for the problem" you mean to create  issue for issue for this  problem?

Comment (jiridanek) [reply .github/workflows/build.yml]: Yes, exactly. So that we don't forget to fix it later.

**Comment by jiridanek** [comment]
It is becoming hard to track what changes you made, between the force pushes. So let me commit this, and post any remaining TODOs as GitHub issues for you.
